kernel/aster-nix/src/process/sync/condvar.rs

```diff
@@ -0,0 +1,192 @@
+                None
+            }
+        };
+        // Wait until the condition becomes true, we're explicitly woken up, or the timeout elapses.
+        self.waiter_counter.lock().fetch_add(1, Ordering::SeqCst);
+        let res = self.pauser.pause_until_or_timeout(cond, &timeout);
+        match res {
+            // OK(): The thread is woken up normally
+            Ok(_) => Ok((lock.lock(), false)),
+            Err(error) => {
+                self.waiter_counter.lock().fetch_sub(1, Ordering::SeqCst);
+                match error.error() {
+                    Errno::EINTR => Err(LockErr::Interrupted((lock.lock(), false))),
+                    Errno::ETIME => Err(LockErr::Timeout((lock.lock(), true))),
+                    _ => Err(LockErr::Unknown((lock.lock(), false))),
+                }
+            }
+        }
+    }
+
+    /// Wait for the condition to become true, or until the timeout elapses, or until the condition is explicitly woken up.
+    /// bool is true if the timeout is reached.
+    pub fn wait_timeout_while<'a, T, F>(
+        &self,
+        mut guard: MutexGuard<'a, T>,
+        timeout: Duration,
+        mut condition: F,
+    ) -> LockResult<(MutexGuard<'a, T>, bool)>
+    where
+        F: FnMut(&mut T) -> bool,
+    {
+        loop {
+            if !condition(&mut *guard) {
+                return Ok((guard, false));
+            }
+            guard = match self.wait_timeout(guard, timeout) {
+                Ok((guard, timeout_flag)) => guard,
+                Err(LockErr::Timeout((guard, timeout_flag))) => {
+                    return Err(LockErr::Timeout((guard, timeout_flag)))
+                }
+                Err(LockErr::Interrupted(guard)) => return Err(LockErr::Interrupted(guard)),
+                Err(LockErr::Unknown(guard)) => return Err(LockErr::Unknown(guard)),
+            }
+        }
+    }
+
+    // Wait for the condition to become true, and until the condition is explicitly woken up or interupted.
+    pub fn wait_while<'a, T, F>(
+        &self,
+        mut guard: MutexGuard<'a, T>,
+        mut condition: F,
+    ) -> LockResult<MutexGuard<'a, T>>
+    where
+        F: FnMut(&mut T) -> bool,
+    {
+        loop {
+            if !condition(&mut *guard) {
+                return Ok(guard);
+            }
+            guard = match self.wait(guard) {
+                Ok(guard) => guard,
+                Err(LockErr::Interrupted(guard)) => return Err(LockErr::Interrupted(guard)),
+                Err(LockErr::Unknown(guard)) => return Err(LockErr::Unknown(guard)),
+                _ => unreachable!(),
+            }
+        }
+    }
+
+    /// Wakes up one task waiting on the condition variable.
+    pub fn notify_one(&self) {
+        let guard = self.waiter_counter.lock_irq_disabled();
+        if guard.load(Ordering::SeqCst) == 0 {
+            return;
+        }
+        self.notify_counter
+            .lock_irq_disabled()
+            .fetch_add(1, Ordering::SeqCst);
```

Comment:
Shoule another check be added here that the `notify_counter` should not exceed `waiter_couner`?

Reply:
Not necessary. `notify_counter` can be even larger when we call `notify`, because although the waiters are woken up, the `wait` still need to read `notify_counter` to return the correct LockResult.